ledger_secure_sdk_sys/build.rs

```diff
@@ -71,26 +71,180 @@ const CCID_FILES: [&str; 9] = [
 ];
 
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
 enum Device {
     NanoS,
     NanoSPlus,
     NanoX,
 }
 
+impl Device {
+    fn to_string(&self) -> String {
+        match self {
+            Device::NanoS => "nanos",
+            Device::NanoSPlus => "nanos2",
+            Device::NanoX => "nanox",
+        }
+        .to_string()
+    }
+}
+
+#[derive(Default)]
+struct SDKInfo {
+    pub bolos_sdk: PathBuf,
+    pub api_level: Option<u32>,
+    pub target_id: String,
+    pub target_name: String,
+    pub c_sdk_name: String,
+    pub c_sdk_hash: String,
+    pub c_sdk_version: String,
+}
+
+impl SDKInfo {
+    pub fn new() -> Self {
+        SDKInfo::default()
+    }
+}
+
+fn retrieve_sdk_info(device: &Device, path: &Path) -> Result<SDKInfo, SDKBuildError> {
+    let mut sdk_info = SDKInfo::new();
+    sdk_info.bolos_sdk = path.to_path_buf();
+    (sdk_info.api_level, sdk_info.c_sdk_name) = retrieve_makefile_infos(&sdk_info.bolos_sdk)?;
+    (sdk_info.target_id, sdk_info.target_name) =
+        retrieve_target_file_infos(&device, &sdk_info.bolos_sdk)?;
+    (sdk_info.c_sdk_hash, sdk_info.c_sdk_version) = retrieve_sdk_git_info(&sdk_info.bolos_sdk);
+    Ok(sdk_info)
+}
+
+fn retrieve_sdk_git_info(bolos_sdk: &Path) -> (String, String) {
+    let c_sdk_hash = match Command::new("git")
+        .arg("-C")
+        .arg(bolos_sdk)
+        .arg("describe")
+        .arg("--always")
+        .arg("--dirty")
+        .arg("--exclude")
+        .arg("*")
+        .arg("--abbrev=40")
+        .output()
+        .ok()
+    {
+        Some(output) => {
+            if output.stdout.is_empty() {
+                "None".to_string()
+            } else {
+                String::from_utf8(output.stdout).unwrap_or("None".to_string())
+            }
+        }
+        None => "None".to_string(),
+    };
+
+    let c_sdk_version = match Command::new("git")
+        .arg("-C")
+        .arg(bolos_sdk)
+        .arg("describe")
+        .arg("--tags")
+        .arg("--exact-match")
+        .arg("--match")
+        .arg("v[0-9]*")
+        .arg("--dirty")
+        .output()
+        .ok()
+    {
+        Some(output) => {
+            if output.stdout.is_empty() {
+                "None".to_string()
+            } else {
+                String::from_utf8(output.stdout).unwrap_or("None".to_string())
+            }
+        }
+        None => "None".to_string(),
+    };
+    (c_sdk_hash, c_sdk_version)
+}
+
+fn retrieve_makefile_infos(bolos_sdk: &Path) -> Result<(Option<u32>, String), SDKBuildError> {
+    let makefile_defines =
+        File::open(bolos_sdk.join("Makefile.defines")).expect("Could not find Makefile.defines");
+    let mut api_level: Option<u32> = None;
+    let mut sdk_name: Option<String> = None;
+    for line in BufReader::new(makefile_defines).lines().flatten() {
+        if let Some(value) = line.split(":=").nth(1).map(str::trim) {
+            if line.contains("API_LEVEL") && api_level.is_none() { 
```

Comment:
```suggestion
            if line.contains("API_LEVEL") && api_level.is_none() {
```
Does fmt check trailing whitespaces?

Reply:
ok CI says it does :)

Reply:
I pushed the formatted code